arangod/Scheduler/WorkStealingThreadPool.cpp

```diff
@@ -0,0 +1,451 @@
+}
+
+void WorkStealingThreadPool::ThreadState::runWork(WorkItem& work) noexcept {
+  incCounter(pool._metrics.jobsDequeued);
+  if (pool._metrics.dequeueTimes) {
+    auto queueTime = std::chrono::duration<double>(
+        std::chrono::steady_clock::now() - work.enqueueTime);
+    auto d =
+        std::chrono::duration_cast<std::chrono::microseconds, double>(queueTime)
+            .count();
+    pool._metrics.dequeueTimes->count(d);
+  }
+  pool.statistics.dequeued.fetch_add(1, std::memory_order_relaxed);
+  try {
+    work.invoke();
+  } catch (...) {
+    LOG_TOPIC("d5fb3", WARN, Logger::FIXME)
+        << "Scheduler just swallowed an exception.";
+  }
+  incCounter(pool._metrics.jobsDone);
+  pool.statistics.done.fetch_add(1, std::memory_order_relaxed);
+}
+
+auto WorkStealingThreadPool::ThreadState::stealWork() noexcept
+    -> std::unique_ptr<WorkItem> {
+  // we start at a random index and iterate through the other threads
+  // using our stepSize. Since the stepSizes are coprimes of the number of
+  // threads this is guaranteed to cover all threads, but different threads will
+  // likely use a different order.
+  auto idx = stealHint.load(std::memory_order_relaxed);
+  if (idx == NoHint) {
+    idx = rng() % pool.numThreads;
+  }
+
+  for (std::size_t i = 0; i < pool.numThreads; ++i) {
+    if (idx != id) {
+      auto& other = pool._threadStates[idx];
+      if (other->pushQueue.load(std::memory_order_relaxed) != nullptr) {
+        auto* item =
+            other->pushQueue.exchange(nullptr, std::memory_order_acquire);
+        if (item) {
+          std::unique_lock guard(mutex);
+          auto res = pushMany(item);
+          // remember this thread so next time we try to steal from it first
+          stealHint.store(idx, std::memory_order_relaxed);
+          return res;
+        }
+      }
+
+      if (other->queueSize.load(std::memory_order_relaxed) > 0) {
+        // stealing is only best effort - try to acquire lock, but don't bother
+        // if we don't get it right away
+        std::unique_lock guard(other->mutex, std::try_to_lock);
+        if (guard.owns_lock() && !other->_queue.empty()) {
+          auto toSteal = std::max<std::size_t>(1, other->_queue.size() / 2);
+
+          auto work = std::move(other->_queue.back());
+          TRI_ASSERT(work != nullptr);
+          other->_queue.pop_back();
+          if (toSteal > 1) {
+            std::unique_lock guard2(mutex);
+            for (std::size_t i = 1; i < toSteal; ++i) {
+              _queue.push_front(std::move(work));
+
+              TRI_ASSERT(other->_queue.size() > 0);
+              work = std::move(other->_queue.back());
+              TRI_ASSERT(work != nullptr);
+              other->_queue.pop_back();
+            }
+            queueSize.store(_queue.size(), std::memory_order_relaxed);
+          }
+          other->queueSize.store(other->_queue.size(),
+                                 std::memory_order_relaxed);
+
```

Comment:
nit: `guard` could be prematurely unlocked here, so that it does not need to wait for the following atomic update to complete.